UCI apply/rollback workflow

jow- opened the pull request:
Switch to rpcd based uci apply/rollback workflow which helps to avoid soft-
bricking devices by requiring an explicit confirmation call after config
apply.

When a user now clicks "Save & Apply", LuCI first issues a call to uci apply
which commits and reloads configuration, then goes into a polling countdown
mode where it repeatedly attempts to call uci confirm.

If the committed configuration is sane, the confirm call will go through and
cancel rpcd's pending rollback timer.

If the configuration change leads to a loss of connectivity (e.g. due to bad
firewall rules or similar), the rollback mechanism will kick in after the
timeout and revert configuration files and pending changes to the pre-apply
state.

In order to cover such rare cases where a lost of connectivity is expected
and desired, the user is offered an "unchecked" apply option after timing
out, which allows committing and applying the changes anyway, without the
extra safety checks.

As a consequence of this change, the luci-reload mechanism is now completely
unsused since rpcd uses ubus config reload signals to reload affected
services, which means that only procd-enabled services will receive proper
reload treatment with the new workflow.

Since this change is rather invasive and only tested with Firefox and my APU so far, I'd like to have some more testing coverage before merging it. Cc @hnyman @dibdot

jow- commented:
You can find a video illustrating the change at http://luci.subsignal.org/~jow/apply.mp4 - it shows different cases of the new mechanism:
1) changing the SSID so that the client looses connection (note how the signal indicator in the upper right corner changes halfway during the timeout and how it pops back once the rollback starts)
2) reverting changes from the uci changelog menu
3) changing a simply setting which does not cause connectivity problems

hnyman commented:
Well, I compiled a new master build r6763-746c590882 with these patches and got mixed results. It is not perfect, yet. Needs finetuning before committing

1)
Change SQM download speed limit
* works ok, if you press Save&Apply
* works ok, if you press Save, then Save&Apply
* fails if you press Save and then visit "Unasved changes" page via the link. From that Unsaved page both Apply and Save&Apply fail always
<img width="589" alt="applysqm_failure" src="https://user-images.githubusercontent.com/7926856/39319204-10f1003c-4989-11e8-83e1-a3c2064bd79d.PNG">


2)
I tried changing wired LAN IP to 192.168.1.4 which should break connectivity and trigger fallback, right?
Results were strange: 
* after a long timeout it first gave error "Apply request failed with status" but the was no further explanation visible
* however, despite that message the config **did change** to 192.168.1.4 and the router was then reachable via SSH console, but not any more from the 192.168.1.1.
<img width="915" alt="applyfallback_reason invisble" src="https://user-images.githubusercontent.com/7926856/39319290-55b7eb36-4989-11e8-9192-9c9377edc114.PNG">

(Note: I am using 150% display scaling in Widnows + 120% in Firefox, so that may impact on visibility of some items, but usually there has been no problem with LuCI.)

jow- commented:
Thanks for testing!

> fails if you press Save and then visit "Unsaved changes" page via the link. From that Unsaved page both Apply and Save&Apply will always fail

This was due to me forgetting to include a modified template in the commit, I ameded it

> after a long timeout it first gave error "Apply request failed with status" but the was no further explanation visible

Did you build include the updated `xhr.js` from master and did you clear your browser cache before? Can you repeat the test with a browser debug panel open to see what requests are happening exactly? Also, did you PC automatically reconnect (received new DHCP) while you performed the test?

hnyman commented:
Your template addition fixed both cases. (And I also cleared the browser cache just in case.)

In any case, now there is the countdown indicator 27...0 sec, then the rollback counter and then the config returned back to normal. Both a normal change and the connectivity-braking change worked as expected.

One thing: the text "Dismiss" on the button is not intuitive. I got it wrong first. 
"Dismiss" just dismisses the warning screen but leaves the pending changes intact, right? The "Revert" button is below there at the normal place, but maybe that should be added to the Dismiss - Force apply row, too.   

As the user has three options, right? 
* Force apply
* Dismiss (and continue editing the config changes for the next try)
* Revert 

I would like to invent a better word for that "Dismiss" but did not yet figure it out.

Ps. I am just wondering if that browser cache clearing will be needed in future when going back and forth between master and 17.01. If really so, that will hurt people updating from 17.01 to master (or 18.05) after this change.

jow- commented:
Maybe it should be “Close | Revert | Apply unchecked” instead?

As for the browser cache issue - normally we append a cache busting string (...?v=git.18...) to each css and js url, so that shouldn’t be a problem for clean builds

hnyman commented:
There might also be some longer explanation of the three choices on the warning textbox.
The choices will not be quite that obvious to the casual user. (as even I understood Dismiss wrongly at first).

Proposal addition:

> The device could not be reached within %d seconds after applying the pending changes, which caused the configuration to be rolled back for safety reasons. If you believe that the configuration changes are correct nonetheless, perform an unchecked configuration apply. _**Alternatively, you can dismiss this warning and edit changes before applying again, or you can revert the changes.**_

jow- commented:
I amended the commit to include your suggested changes.
![revised](https://user-images.githubusercontent.com/2528802/39346215-3f59c9f4-49ed-11e8-9bf5-4b1e66877e95.png)

hnyman commented:
Looks nice and clear.

dibdot commented:
@jow- just tested your revised change on a quite slow travel router this morning with latest LuCI Master (git-18.116.27617-aa6c971) . In general it works quite fine - just two nitpicks:
- the usual "Save & Apply" workflow always ends with the visible message "There are no changes to apply" which is quite confusing, at least for me.
- it's still possible to delete your LAN interface ... without rollback ;-)

Just a first impression, will do more testing during this weekend. Thanks!

jow- commented:
@dibot - thanks for giving it a try!

- Does the “No changes to apply” message always occur or only if you hit “Save and apply” without doing actual changes?

- It is true that one still can delete interfaces and networks, I will rework those overview pages to be real cbi maps later, so that they benefit from the same mechanisms

dibdot commented:
the message always occur, e.g. whenever I change the design or ntp server settings etc.

hnyman commented:
I noticed the same. The message briefly flashes visible and then quickly disappears.

jow- commented:
Allright, will try to reproduce

jow- commented:
Does that flashing happen before the countdown? Any chance that you can watch the request flow in the debug console while this happens?

hnyman commented:
It happens right at the end of the workflow, just before the page toggles from appyl-mode to the normal settings page.

Easy example: 
* toggle a collectd plugin enabled/disabled (I used Uptime for testing)
* press "Save&Apply"
* First comes text for 1 sec: "Starting configuration apply..."
* Then for 1 sec: "There are no change to apply"
* Then the process completes

Countdown was not visible at all.

If I press "Save" and then go to the "Unsaved changes" page, and "Save & Apply", I get the countdown and "...has been applied".

hnyman commented:
Additional note:
the same 1 sec: "Starting configuration apply..." & 1 sec: "There are no changes to apply" comes, if you select "Revert" on the "Unchanged changes"  screen.

jow- commented:
Hmm thats odd. A 204 return value to the `apply_rollback` request can only happen when there are no uncomitted changes in the per session save directory... Can you run `ubus monitor` while running the Save & Apply? Also, if you run just Save, is there a `/var/run/rpcd/uci-*` directory containing the changes?

Seems there is some race condition of some sort

hnyman commented:
The setting in /etc/config/luci_statistics actual does change, so the transaction is completed. I checked the file before and after.

After Save, there is the file in /var/run/rpcd/uci-*
And it disappears after Apply&Save.

To me it looks like the transaction completed almost immediately, maybe before the wait-polling mechanism kicks in? So there are no changes any more.

New example:
ubus log: https://pastebin.com/CaxDcGMB

![image](https://user-images.githubusercontent.com/7926856/39397432-a268baee-4b07-11e8-8a0b-d91020130c1d.png)

jow- commented:
Thanks for the input @hnyman - I was able to reproduce it on a completely fresh build. It turned out to be a missing change to `cbi.lua` in the rollback commit:

```diff
diff --git a/modules/luci-base/luasrc/cbi.lua b/modules/luci-base/luasrc/cbi.lua
index b16c91ea2..5ec7b468a 100644
--- a/modules/luci-base/luasrc/cbi.lua
+++ b/modules/luci-base/luasrc/cbi.lua
@@ -395,8 +395,10 @@ function Map.parse(self, readinput, ...)
                self:_run_hooks("on_after_save")
                if (not self.proceed and self.flow.autoapply) or luci.http.formvalue("cbi.apply") then
                        self:_run_hooks("on_before_commit")
-                       for i, config in ipairs(self.parsechain) do
-                               self.uci:commit(config)
+                       if self.apply_on_parse then
+                               for i, config in ipairs(self.parsechain) do
+                                       self.uci:commit(config)
+                               end
                        end
                        self:_run_hooks("on_commit", "on_after_commit", "on_before_apply")
                        if self.apply_on_parse then

```

I amended the commit accordingly.

hnyman commented:
hmmm....
was that right?

It seems to break the whole thing. Now the countdown starts, but never notices success, then it goes to rollback, and never stops waiting for the rollback completion.

hnyman commented:
If it works ok for @dibdot then it is likely ok. I will re-compile the whole stuff and reflash. (Earlier I just recompiled luci-base and installed that.)

hnyman commented:
Sorry, but recompiling and flashing did not help.

Interestingly, Firefox 60.0b16 gets stuck in the loop, while IE11 gives an error in debugger:

Firefox:
![image](https://user-images.githubusercontent.com/7926856/39400547-ff6c44da-4b3a-11e8-93f0-af3d7d2f1812.png)

IE:
![image](https://user-images.githubusercontent.com/7926856/39400551-28506e58-4b3b-11e8-86d0-e2994b1d7cc2.png)

![image](https://user-images.githubusercontent.com/7926856/39400552-30f2e2fc-4b3b-11e8-8f26-b64bbc8ed206.png)

jow- commented:
Interesting - can you check if the following diff fixes the IE problem?

```diff
diff --git a/modules/luci-base/htdocs/luci-static/resources/xhr.js b/modules/luci-base/htdocs/luci-static/resources/xhr.js
index 91dcf3fef..de4476cdd 100644
--- a/modules/luci-base/htdocs/luci-static/resources/xhr.js
+++ b/modules/luci-base/htdocs/luci-static/resources/xhr.js
@@ -54,11 +54,11 @@ XHR = function()
                        else
                                url += '?' + code;
 
+               xhr.open('GET', url, true);
+
                if (!isNaN(timeout))
                        xhr.timeout = timeout;
 
-               xhr.open('GET', url, true);
-
                xhr.onreadystatechange = function()
                {
                        if (xhr.readyState == 4) {
@@ -92,10 +92,11 @@ XHR = function()
                                callback(xhr);
                }
 
+               xhr.open('POST', url, true);
+
                if (!isNaN(timeout))
                        xhr.timeout = timeout;
 
-               xhr.open('POST', url, true);
                xhr.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
                xhr.send(code);
        }

```

hnyman commented:
Yes, that fixed IE11 (see below), but no impact on Firefox 60

![image](https://user-images.githubusercontent.com/7926856/39400662-946f36a8-4b3d-11e8-9312-3dd9f4da697e.png)

hnyman commented:
Hmm.
Tested with Chrome and it works (with the IE-triggered change).
Tested with Firefox in safe mode without plugins, and it works.

Apparently it is an effect from a Firefox plugin/extension :-(
I don't even have that many plugins in use (adblock plus, ghostery, zoompageWE)

hnyman commented:
So far the latest changes (1491ddd and 5cc8f8a) have worked ok in my primary router (even with the Firefox plugins). Strange, as my other router had the same changes edited by hand and has given varying results with Firefox as I told above.

Will test again after recompiling the whole firmware also for that.

hnyman commented:
Strange. 
The primary router at 192.168.1.1 is ok with these changes, while the other router at 192.168.1.2 (connected via the main one at .1) gives timeouts. Same Firefox, same plugins. Also IE11 seems to occasionally give timeout with .2

dibdot commented:
More oddities from my side: Tested this morning with latest patchset, connected directly via ethernet cable). Started with a fresh wireless config (rm /etc/config/wireless, wifi config), enter LuCI wireless tab, change the SSID and hit "Save & Apply"  - changes will only be saved but not applied. 
![setup_ap1](https://user-images.githubusercontent.com/8061346/39403729-128fff32-4b83-11e8-9fa1-a0842ce38377.png)

That's always reproducible. 

If I click "Save & Apply" again changes will be applied ... or I'll get a loop, but that's not always reproducible.
![setup_ap2](https://user-images.githubusercontent.com/8061346/39403748-8a35975e-4b83-11e8-83a9-014895f6cc90.png)

Firefox 59.0.2/Linux without any plugins

jow- commented:
Thanks for the feedback - will try some more test on a TP-Link today.

hnyman commented:
Now it seems to pick the config changes ok. I could not manage to get it to fail. The simple config change also always got confirmed quickly, so the changed timing has improved communication.

But negatively, it is now possible to make a soft-brick change that does not get rolled back:

I tried setting IP to 192.168.1.4 and that change went through. Browser got waiting for rollback, but that never materialised. I finally SSHed in and changed IP from .4 back to .2

hnyman commented:
I re-compiled whole firmwares for both ar71xx and ipq806x and this time the LuCI rollback worked as expected, valid changes went through quickly and invalid ones saw rollback as expected.

(but the wireshark log I sent earlier shows that sometimes apparently the rollback may be left undone. No idea why that happened then, but the same change now did rolled properly back.)

jow- commented:
Great to hear. Try increasing the 2000ms grace wait interval until the rollback confirm kicks in to something like 5000ms, maybe the confirm happens too early now, before the services even got restarted.

Regards,
Jo

jeffsf commented:
A configurable timeout (even if not exposed in LuCI) would be helpful for a handful of situations.

As a specific example, most of my APs only have connectivity over tunnels on top of 802.11s. Most have several virtual APs/devices on both bands. As a result, it can take 10-20 seconds from a configuration change before all the APs/devices are created and come up, then the mesh needs to connect, then the tunnels, before connectivity is reestablished to the device.

jow- commented:
> A configurable timeout (even if not exposed in LuCI) would be helpful for a handful of situations

Add the following to /etc/config/luci:

```
config internal apply
    option timeout 60  # in seconds
```

hnyman commented:
Looks like a regression: the applied config changes do not cause a process restart at least for some apps, so the new config is not actually applied. At least SQM and luci-statistics seems affected.

I just tested and
*    with plain r6755 (without LuCI rollback) both SQM and luci_statistics config changes kicked off a process restart
*    with r6787+rollback the process restart does not happen

hnyman commented:
@dibdot  @jow- 
I think that luci-app-adblock misbehaves with the newest rollback code:
* There is just "Save & apply" button, but that just saves.
* That button saves the config change and triggers adblock reload, but does not actually do the Apply action before triggering reload. Changes remain saved and are visible on the Unsaved changes page, where they can be then applied.

I gusss that this might be due to the adblock-app specific reload-override?
https://github.com/openwrt/luci/blob/master/applications/luci-app-adblock/luasrc/model/cbi/adblock/overview_tab.lua#L28
```
function m.on_after_commit(self)
	luci.sys.call("/etc/init.d/adblock reload >/dev/null 2>&1")
	luci.http.redirect(luci.dispatcher.build_url("admin", "services", "adblock"))
end
```
Removing that redirect seems to trigger the new apply process. Maybe the redirect in that function is too early and the current rollback code (with apply) gets never activated.

There might be other apps with similar redirects, some of which might be needed.

rwalli commented:
Hello!

Sorry for double-post, but i think here is the right place:

Wanted to try out this failsafe mechanism but could not get it to work.
Do you need to activate any special options in menuconfig?
(i only activated luci)
patch went through without errors…
Robert

hnyman commented:
> Do you need to activate any special options in menuconfig?

No special actions needed.
Apply the three patches (the three commits here) and fully recompile and install LuCI (at least luci-base and luci-mod-admin-full). After installing them, you should reboot the router and probably also clear the browser cache.

rwalli commented:
On 1 May 2018, at 20:50, Hannu Nyman <notifications@github.com> wrote:Do you need to activate any special options in menuconfig?
No special actions needed.
Apply the three patches (the three commits here) and fully recompile and install LuCI (at least luci-base and luci-mod-admin-full). After installing them, you should reboot the router and probably also clear the browser cache.I applied this patch: https://patch-diff.githubusercontent.com/raw/openwrt/luci/pull/1769.patchThe wrong one?robert

hnyman commented:
Looks right in principle, as it has all three commits in a row. But check that they all get applied properly. (I am more used to download the individual commits as patches. I have never tried that combined patch of a pull request)

rwalli commented:
> On 1 May 2018, at 21:05, Hannu Nyman <notifications@github.com> wrote:
> 
> Looks right in principle, as it has all three commits in a row. But check that they all get applied properly. (I am more used to download the individual commits as patches. I have never tried that combined patch of a pull request)
> 
> 


I’m sure that the patches did not raise any errors yesterday.
I just removed the whole lede directory, and started from scratch but now not all hunk’s succeeded.
Could it be the changed source because of 2 pull-requests in the last 24h?

Robert

jow- commented:
> Looks like a regression: the applied config changes do not cause a process restart at least for some apps, so the new config is not actually applied. At least SQM and luci-statistics seem affected.

Yes, both packages need their init scripts converted to procd ones in order to work with the new apply mechanism.

jow- commented:
> I gusss that this might be due to the adblock-app specific reload-override?
https://github.com/openwrt/luci/blob/master/applications/luci-app-adblock/luasrc/model/cbi/adblock/overview_tab.lua#L28

Yes, that code needs to be replaced with a simple `m.redirect = luci.dispatcher.build_url("admin", "services", "adblock")`

jow- commented:
Rebased the series on top of LuCI master and introduced an `ucitrack` service which takes care of handling non-procd services.

rwalli commented:
It's a while back i build openwrt...
That must be really a stupid mistake!

Double checked everything:

- removed privious checkout
- switched from "https://git.lede-project.org/source.git" to "https://github.com/openwrt/openwrt.git"
- "patch --verbose" reported all hunks successful (attached)
- examine samples of files if the patch is really there
- renamed the xxx-sysupgrade.bin file to avoid flashing the wrong file

No success!
Any suggestions?

Robert

[patch_output.txt](https://github.com/openwrt/luci/files/1966568/patch_output.txt)

jow- commented:
@rwalli - what is the problem?

rwalli commented:
Luci behaves like normal.

hnyman commented:
@jow- 
Thanks for fixing sqm and statistics. They now work ok.
Other apps with redirects naturally still misbehave, like @dibdot 's adblock.


Regarding the regression I mentioned earlier, that IP change to 192.168.1.4 did not rollback properly:

> > I tried setting IP to 192.168.1.4 and that change went through. Browser got waiting for rollback, but that never materialised. I finally SSHed in and changed IP from .4 back to .2
>
> Try increasing the 2000ms grace wait interval until the rollback confirm kicks in to something like 5000ms, maybe the confirm happens too early now, before the services even got restarted.

Your suggestion to change the 2000 ms initial wait time was right. I increased it to 4000 ms and this time the rollback worked as expected. (with current 2000 ms the current code fails to notice the need for the rollback)

root@OpenWrt:~# diff -u /rom/usr/lib/lua/luci/view/cbi/apply_widget.htm /usr/lib/lua/luci/view/cbi/apply_widget.htm
```
--- /rom/usr/lib/lua/luci/view/cbi/apply_widget.htm
+++ /usr/lib/lua/luci/view/cbi/apply_widget.htm
@@ -133,7 +133,7 @@
                                var deadline = Date.now() + uci_apply_timeout * 1000;
                                window.setTimeout(function() {
                                        uci_confirm(checked, deadline);
-                               }, 2000);
+                               }, 4000);
                        }
                        else if (checked && r.status === 204) {
                                stat.innerHTML = '<%:There are no changes to apply.%>';

```

dibdot commented:
@hnyman @jow I will fix adblock & travelmate in the next few days, still I haven't fully understand the new redirect behaviour ... currently I'm unable to force a page reload (that's what I have done before with the http.redirect stuff).

@jow- latest patchset works fine for me, only a visual glitch:
before workflow (section alignment is OK)
![before](https://user-images.githubusercontent.com/8061346/39541696-cd1fc5ec-4e46-11e8-9add-7e8651965d04.png)
after workflow (section alignment is "collapsed")
![after](https://user-images.githubusercontent.com/8061346/39541800-12b95258-4e47-11e8-896e-0a011cce52dc.png)

jow- commented:
> I will fix adblock & travelmate in the next few days, still I haven't fully understand the new redirect behaviour ... currently I'm unable to force a page reload (that's what I have done before with the http.redirect stuff).

Best is to simply set `m.redirect = desired_target_url` that should take care of redirecting after apply.

> latest patchset works fine for me, only a visual glitch:
Yeah, thats a bug with the material theme, it has quite some issues I plan to address later.

jow- commented:
> Other apps with redirects naturally still misbehave, like @dibdot 's adblock.
Even after installing and starting the new `/etc/init.d/ucitrack` init?

hnyman commented:
At least with the old redirect line, adblock jumps over the rollback screen, I think.
I changed the redirect line along your suggestion, but there are still somehow double items in log.  (I left the reload command into on_after_commit ). Something for dibdot to ponder...
```
Wed May  2 21:54:57 2018 user.info adblock-[3.5.1]: start adblock processing (reload)
Wed May  2 21:55:00 2018 daemon.info dnsmasq[3286]: read /etc/hosts - 4 addresses
...
Wed May  2 21:55:00 2018 daemon.info dnsmasq[3286]: using nameserver 2001:14b8:1000::2#53
Wed May  2 21:55:00 2018 daemon.info dnsmasq[3286]: using 5146 more local addresses
Wed May  2 21:55:00 2018 user.info adblock-[3.5.1]: blocklist with overall 5147 domains loaded successfully (Netgear Nighthawk X4S R7800, OpenWrt SNAPSHOT r6807-58f7b5b96c)
Wed May  2 21:55:00 2018 user.info adblock-[3.5.1]: start adblock processing (reload)
Wed May  2 21:55:04 2018 user.info adblock-[3.5.1]: blocklist with overall 5147 domains loaded successfully (Netgear Nighthawk X4S R7800, OpenWrt SNAPSHOT r6807-58f7b5b96c)

```

dibdot commented:
@jow- 
> Best is to simply set m.redirect = desired_target_url that should take care of redirecting after apply.

That does not work for me. Whenever I add the m.redirect, the rendered page get a (unwanted) "Back to Overview" control but no reload/redirect after "Save & Apply", e.g.:

![m redirect](https://user-images.githubusercontent.com/8061346/39545215-c7d82b2e-4e50-11e8-99f9-accd4fbfb32d.png)

jow- commented:
@dibdot - travelmate uses SimpleForm instances, those do not use the rollback at all.

rwalli commented:
Hello!

I guess the patching worked properly.
(have reviewed many patched files on the device)

But I'v tried only firewall rules because @jow- mentioned in the first post.
Luci behaved here as normal and locked me out without rollback warning or log entries.

Today i tried to change ip-address and the rollback warning worked.
When deleting the LAN-Interface, the device locked me out before i could click on the "Save & Apply" button.

Robert

jow- commented:
@rwalli - deleting the lan interface not rolling back is expected for now, the delete still action bypasses the normal mechanisms and need to be ported over after this PR has been merged.

As for the firewall lockout still happening; might be caused by conntrack not getting flushed, will check it later.

jow- commented:
@hnyman - patchset rebased, squashed and updated. All timings are now configurable in `/etc/config/luci`:

```
config internal 'apply'
	option rollback '30'
	option holdoff '4'
	option timeout '3'
	option display '1.5'
```

rwalli commented:
> As for the firewall lockout still happening; might be caused by conntrack not getting flushed
seems to be right, when i log in with ssh the connection does not disconnect.

hnyman commented:
@jow- 
I am still puzzled with the SQM behaviour. Even with the newest rollback code, after a clean boot, the changes are applied ok to /etc/config/sqm but that does not trigger reloading the service on the router. But if I restart SQM once from console, then the same kind changes made in LuCI start triggering the service reload. 

Not quite sure how to debug that. procd problem? or something to do with the init script?

jow- commented:
@hnyman - this indeed looks like some bug in the sqm / procd ballpark, I'll look into that separately from this issue.

hnyman commented:
@jow- 
Regarding SQM and rollback: I think that it is somehow still related to this rollback code, as the same SQM and procd version behave differently with/without rollback code.

**without** LuCI rollback code
* build r6835-e495a05069, making SQM config changes in LuCI after boot--> triggers SQM restart

**with** LuCI rollback code
* build r6808-3a456683e5 (yesterday), making SQM config changes in LuCI after boot--> no  SQM restart
* build r6808-3a456683e5 (yesterday), making SQM config changes in LuCI after SQM has been once restarted in console  --> triggers SQM restart

jow- commented:
Yeah, could be. Old behaviour triggered an explicit `/etc/init.d/sqm reload` through luci-reload (ucitrack), current code relies on `ubus call service event '{ "type": "config.change", "data": { "package": "sqm" } }` to trigger service restart.

hnyman commented:
>  The bug is in the procd sqm init script. Due to redeclaration of boot() which directly invokes sqm instead of using the normal service start up procedure, no sqm service is registered with procd. 

Interesting.

I just wonder if there are other apps with similar fault in init script. 
"boot" being not enough and only "start" making the service registration is news to me (and apparently also for you), so it it is quite possible that there are other similar ones.  At  quick scan in the packages feed, I spotted one possibility: "acme" uses procd and has non-standard boot function.

jow- commented:
Yeah, quite likely.

hnyman commented:
@jow- 
Shouldn't the xhr IE11 fix be merged separately, before the 18.06 branching? 
I get the same "InvalidStateError" in IE11 debugger for a clean LuCI.

I think that the fix has nothing to do with the rollback stuff. The earlier XHR change has already been committed and that is just a fix for that.

jow- commented:
I agree, fix committed separately.

tohojo commented:
Is there a way to register the service with procd without calling the start function? I merged the change to sqm-scripts, but for ACME it's not actually a good idea to run the same command at boot...

jow- commented:
> Is there a way to register the service with procd without calling the start function? I merged the change to sqm-scripts, but for ACME it's not actually a good idea to run the same command at boot...

@tohojo - I've been thinking about this too and couldn't come to a satisfactory solution... in general the cleanest way for now seems to be to handle the "am I starting at boot?" case in the application itself and to communicate the on-boot state out of band, e.g. via a stamp file.

Something like this:
```!diff
diff --git a/net/acme/files/acme.init b/net/acme/files/acme.init
index 6de12b7d..a3d686dc 100644
--- a/net/acme/files/acme.init
+++ b/net/acme/files/acme.init
@@ -25,7 +25,8 @@ stop_service() {
 }
 
 boot() {
-    $SCRIPT check_cron
+    touch "/var/run/acme_boot"
+    start
 }
 
 service_triggers()
diff --git a/net/acme/files/run.sh b/net/acme/files/run.sh
index a3a89ab9..368de26d 100644
--- a/net/acme/files/run.sh
+++ b/net/acme/files/run.sh
@@ -234,6 +234,7 @@ load_vars()
 
 check_cron
 [ -n "$CHECK_CRON" ] && exit 0
+[ -e "/var/run/acme_boot" ] && rm -f "/var/run/acme_boot" && exit 0
 
 config_load acme
 config_foreach load_vars acme
```

It is not the most elegant solution; I suppose the best would be if procd could expose the state via an environment variable or something, but so far we do not have such a facility.

dibdot commented:
In adblock I'm using this modified boot function:

```
boot()
{
    adb_boot=1
    rc_procd start_service
}
[...]
start_service()
{
    if [ $("${adb_init}" enabled; printf "%u" ${?}) -eq 0 ]
    then
        if [ -n "${adb_boot}" ]
        then
            local trigger="$(uci_get adblock global adb_trigger)"
            if [ "${trigger}" != "timed" ]
            then
                return 0
            fi
        fi
[...]
```

This means usually only the trigger will be initialized on boot - works quite fine.

EricLuehrsen commented:
@jow- @dibdot - luci-app-unbound has the same save-apply regression side effects.

dibdot commented:
@EricLuehrsen see #1783, should work, but untested.

dibdot commented:
@hnyman @jow- all workflow related changes in unbound, travelmate & adblock are successfully re-tested with new & current workflow and applied to master.

hnyman commented:
@jow- 
are you going to merge this into master now that 18.06 has been branched off?

I feel that it works ok for most parts, and wider testing/feedback is needed for finding out the remaining problems.

guidosarducci commented on modules/luci-base/root/etc/init.d/ucitrack:
```diff
@@ -0,0 +1,57 @@
+		esac
+	fi
+
+	for affected in $affects; do
```
I think this code reverses `$config` and `$affected`, as well as ignores whether the `$affected` init uses `procd`.

e.g. it would be simpler and more correct to do something like:
```
	for affected in $affects; do
		register_init "$config" "/etc/init.d/$affected" "reload"
	done
```
but even that ignores the case of `$affected` having an `exec` entry in `ucitrack`.

Perhaps restructure rewrite `register_init()` to be `register_init_or_exec()`, and then use the latter to iterate over the combined list of `$config` and `$affects`.

Hopefully I didn't miss something obvious. :-)

jow- replied:
The logic is intentionally reversed; procd already takes care of reloading services whose own configs changed and LuCI will emit config.change notifications for those. This code here sets up the reverse dependencies procd does not know about, meaning we will reload *our* service if any of the potentially affected configs changed.

We also assume that affected entries are either procd enabled or covered by ucitrack as well (end thus have a dummy service entry reacting on reload signals).

guidosarducci replied:
I'm agreeing with what you say, just thinking the implementation doesn't line up. Let's take a concrete example of `/etc/config/ucitrack`:
```
config firewall
        option init 'firewall'
        list affects 'sqm'
config sqm
        option init 'sqm'
```
This means that `firewall` config changes should cause `sqm` service to reload (and rightly so) in the old "luci-reload" world.

In the new world of "procd + luci rollback", `/etc/init.d/{firewall,sqm}` are each procd-enabled and add a trigger to self-reload on changes to their own config. So far so good. But why then does `/etc/init.d/ucitrack` handle the `list affects 'sqm'` by issuing the procd command:
```
        procd_add_config_trigger "config.change" "sqm" \
            ubus call service event \
            '{"type":"config.change","data":{"package":"firewall"}}'
```
The `firewall` service *should not* be reloaded on a `sqm` config change. This is opposite to the `/etc/config/ucitrack` config above.

Help me understand why you think this opposite behaviour is correct. I must be missing something, or I'm up too late.

As an aside, my own `sqm-scripts` fork also includes a `procd_add_reload_trigger "firewall"` precisely to remove any `ucitrack` dependency. Isn't that the right approach? That's why I thought it reasonable to check if an "affected" init script is not procd-enabled, and only then issue the correct "cross-service" config trigger.

jow- replied:
> The firewall service should not be reloaded on a sqm config change. This is opposite to the /etc/config/ucitrack config above.

The firewall should get reloaded. The old luci-reload (and luci.model.uci.apply()) implementation issued an init reload for each affected package including itself. So if any of sqm or firewall got changed, both sqm and firewall got reloaded. The above takes care of doing that in the reverse direction by broadcasting a change event and letting procd sort out the details.

> As an aside, my own sqm-scripts fork also includes a procd_add_reload_trigger "firewall" precisely to remove any ucitrack dependency. Isn't that the right approach?

It is.

> That's why I thought it reasonable to check if an "affected" init script is not procd-enabled, and only then issue the correct "cross-service" config trigger.

If an init script is procd-enabled, the extraneous config.change event is harmless. Procd will a) caolesce exevents and b) no-op the change event if the calculated desired runtime change is identical to the effective runtime state (it compares command lines and keeps checksums of all related configurations).

jow- replied:
Please also keep in mind that this is a transitional mechanism. I plan to fade out ucitrack entirely within the next few months. No package should rely on it anymore. The sole reason I did that is to keep package not converted to procd somewhat working yet.

guidosarducci replied:
Hmmm. I had set aside the problems I saw on the expectation that `ucitrack` would be quickly phased out. However, it's now been more than 2 years since this transitional mechanism was to be retired, and the same issues persist. What are the remaining obstacles to fade out `ucitrack` entirely as planned?

tganeshg commented:
@jow- 

I have same problem and It was working with previous versions of LuCI:

In my cbi I have an on_after_commit() which calls a script via sys.call() to convert cbi config file to json. At the time the script gets called , it seems before commit it called.

on_after_apply() not at calling..

Do you have any solution/workaround to handle the new behaviour of LuCI?

Thanks in advance

fhqihgu commented:
How to close the rollback function simply, click a button or edit a config e.g.
And, I think it's Inconvenient，suggest to close it forever.
